src/fides/api/ops/util/consent_util.py

```diff
@@ -48,19 +56,105 @@ def should_opt_in_to_service(
+        for pref in relevant_preferences
+        if pref.preference == preference_to_propagate
+    ]
+
+    # Return whether we should opt in, and the filtered preferences so we can update those for consent reporting
+    return preference_to_propagate == UserConsentPreference.opt_in, filtered_preferences
+
+
+def cache_initial_status_and_identities_for_consent_reporting(
+    privacy_request: PrivacyRequest,
+    connection_config: ConnectionConfig,
+    relevant_preferences: List[PrivacyPreferenceHistory],
+    relevant_user_identities: Dict[str, Any],
+) -> None:
+    """Add a pending system status and cache relevant identities on the applicable PrivacyPreferenceHistory
+    records for consent reporting.
+
+    Preferences that aren't relevant for the given system/connector are given a skipped status.
+
+    Typically used when *some* but not all privacy preferences are relevant.  Otherwise,
+    other methods just mark all the preferences as skipped.
+    """
+    for pref in privacy_request.privacy_preferences:
+        if pref in relevant_preferences:
+            pref.update_secondary_user_ids(relevant_user_identities)
+            pref.cache_system_status(
+                connection_config.system_key, ExecutionLogStatus.pending
+            )
+        else:
+            pref.cache_system_status(
+                connection_config.system_key, ExecutionLogStatus.skipped
+            )
+
+
+def add_complete_system_status_for_consent_reporting(
+    privacy_request: PrivacyRequest,
+    connection_config: ConnectionConfig,
+) -> None:
+    """Cache a complete system status for consent reporting on just the subset
+    of preferences that were deemed relevant for the connector on failure
+
+    Deeming them relevant if they already had a "pending" log added to them.
+    """
+    for pref in privacy_request.privacy_preferences:
+        if (
+            pref.affected_system_status
+            and pref.affected_system_status.get(connection_config.system_key)
+            == ExecutionLogStatus.pending.value
+        ):
+            pref.cache_system_status(
+                connection_config.system_key,
+                ExecutionLogStatus.complete,
+            )
+
+
+def add_errored_system_status_for_consent_reporting(
+    privacy_request: PrivacyRequest,
+    connection_config: ConnectionConfig,
+) -> None:
+    """Cache an errored system status for consent reporting on just the subset
+    of preferences that were deemed relevant for the connector on failure
+
+    Deeming them relevant if they already had a "pending" log added to them.
+    """
+    for pref in privacy_request.privacy_preferences:
+        if (
+            pref.affected_system_status
+            and pref.affected_system_status.get(connection_config.system_key)
+            == ExecutionLogStatus.pending.value
+        ):
+            pref.cache_system_status(
+                connection_config.system_key,
+                ExecutionLogStatus.error,
+            )
```

Comment:
These methods are similar shared between saas and email consent connectors to update affected system status with a complete or errored status.  This is only for preferences that were relevant for the given system - the others will have a skipped status.